### PR TITLE
add "fork" method to DenoPermissions, PermissionState

### PR DESCRIPTION
**Pull request by humancalico**
UPDATE: This is outdated now. Since `DenoPermission` struct was renamed to `Permissions`
For #4779

### CODE REVIEW

File: cli/permissions.rs [hunk 1]
```diff
@@ -288,6 +288,109 @@ impl DenoPermissions {
...
+    allow_write: bool,
+    write_whitelist: HashSet<PathBuf>,
+    allow_net: bool,
+    net_whitelist: HashSet<String>,
+    allow_env: bool,
+    allow_run: bool,
+    allow_plugin: bool,
+    allow_hrtime: bool,
+  ) -> Result<DenoPermissions, OpError> {
+    if allow_read && self.allow_read == PermissionState::Deny {
```
Comment (bartlomieju) [inline cli/permissions.rs:306]: How about adding `fork` method to `PermissionState`? `self.allow_read.fork(allow_read)?` - then you can encapsulate check and returning error in a single function

File: cli/permissions.rs [hunk 2]
```diff
@@ -288,6 +288,109 @@ impl DenoPermissions {
       n => Err(OpError::other(format!("No such permission name: {}", n))),
     }
   }
+
+  #[allow(clippy::too_many_arguments)]
```
Comment (bartlomieju) [inline cli/permissions.rs:292]: Ok

File: cli/permissions.rs [hunk 3]
```diff
@@ -288,6 +288,109 @@ impl DenoPermissions {
...
+      Err(OpError::permission_denied(
+        "Permission passed. As parent permissions have arguments escalate parent permissions. As parent permissions have `allow_hrtime: false`"
+          .to_string(),
+      ))
+    } else if !(read_whitelist.is_subset(&self.read_whitelist)) {
+      Err(OpError::permission_denied(format!(
+        "Permission passed as arguments escalate parent permissions has only {:?} in `read_whitelist`",
+        self.read_whitelist
+      )))
+    } else if !(write_whitelist.is_subset(&self.write_whitelist)) {
```
Comment (bartlomieju) [inline cli/permissions.rs:341]: Nice

File: cli/permissions.rs [hunk 4]
```diff
@@ -288,6 +288,109 @@ impl DenoPermissions {
...
+          PermissionState::Deny
+        },
+        allow_hrtime: if allow_hrtime {
+          PermissionState::Allow
+        } else {
+          PermissionState::Deny
+        },
+      })
+    }
+  }
```
Comment (bartlomieju) [inline cli/permissions.rs:393]: Please add unit tests

File: cli/permissions.rs (hunk 3 above)
Comment (humancalico) [reply cli/permissions.rs:341]: Oops didn't see that :grimacing:

**Pull request closed without merging**